Mathlib/Topology/Homotopy/HomotopyGroup.lean

```diff
@@ -0,0 +1,286 @@
+    Function.funext_iff, Function.funext_iff, Fin.forall_fin_one, Fin.forall_fin_one]
+  rfl
+
+end Cube
+
+/--
+The `n`-dimensional generalized loops; functions `I^n → X` that send the boundary to the base point.
+-/
+structure GenLoop (n : ℕ) (x : X) extends C(I^ n, X) where
+  boundary : ∀ y ∈ Cube.boundary n, toFun y = x
+#align gen_loop GenLoop
+
+namespace GenLoop
+
+instance funLike : FunLike (GenLoop n x) (I^ n) fun _ => X where
+  coe f := f.1
+  coe_injective' := fun ⟨⟨f, _⟩, _⟩ ⟨⟨g, _⟩, _⟩ _ => by congr
+#align gen_loop.fun_like GenLoop.funLike
+
+@[ext]
+theorem ext (f g : GenLoop n x) (H : ∀ y, f y = g y) : f = g :=
+  FunLike.ext f g H
+#align gen_loop.ext GenLoop.ext
+
+@[simp]
+theorem mk_apply (f : C(I^ n, X)) (H y) : (⟨f, H⟩ : GenLoop n x) y = f y :=
+  rfl
+#align gen_loop.mk_apply GenLoop.mk_apply
+
+@[simp] -- porting note: new lemma
+theorem toContinuousMap_apply (f : GenLoop n x) (y) : f.toContinuousMap y = f y := rfl
+
+initialize_simps_projections GenLoop (toFun → apply)
+
+/-- The constant `GenLoop` at `x`.
+-/
+@[simps!] -- porting note: new attr
+def const : GenLoop n x :=
+  ⟨ContinuousMap.const _ x, fun _ _ => rfl⟩
+#align gen_loop.const GenLoop.const
+
+instance inhabited : Inhabited (GenLoop n x) where default := const
+#align gen_loop.inhabited GenLoop.inhabited
+
+/-- Restrict a `GenLoop n x` with `n ≠ 0` to the diagonal of the cube. -/
+@[simps!]
+def diagonal [NeZero n] (f : GenLoop n x) : Path x x where
+  toContinuousMap := f.comp <| .constPi _
+  source' := f.boundary _ Cube.zero_mem_boundary
+  target' := f.boundary _ Cube.one_mem_boundary
+
+/-- The "homotopy relative to boundary" relation between `GenLoop`s.
+-/
+def Homotopic (f g : GenLoop n x) : Prop :=
+  f.toContinuousMap.HomotopicRel g.toContinuousMap (Cube.boundary n)
+#align gen_loop.homotopic GenLoop.Homotopic
+
+namespace Homotopic
+
+variable {f g h : GenLoop n x}
+
+@[refl]
+theorem refl (f : GenLoop n x) : Homotopic f f :=
+  ContinuousMap.HomotopicRel.refl _
+#align gen_loop.homotopic.refl GenLoop.Homotopic.refl
+
+@[symm]
+nonrec theorem symm (H : f.Homotopic g) : g.Homotopic f :=
+  H.symm
+#align gen_loop.homotopic.symm GenLoop.Homotopic.symm
+
+@[trans]
+nonrec theorem trans (H0 : f.Homotopic g) (H1 : g.Homotopic h) : f.Homotopic h :=
+  H0.trans H1
+#align gen_loop.homotopic.trans GenLoop.Homotopic.trans
+
+theorem equiv : Equivalence (@Homotopic X _ n x) :=
+  ⟨Homotopic.refl, Homotopic.symm, Homotopic.trans⟩
+#align gen_loop.homotopic.equiv GenLoop.Homotopic.equiv
+
+instance setoid (n : ℕ) (x : X) : Setoid (GenLoop n x) :=
+  ⟨Homotopic, equiv⟩
+#align gen_loop.homotopic.setoid GenLoop.Homotopic.setoid
+
+protected theorem diagonal [NeZero n] (H : f.Homotopic g) :
+    f.diagonal.Homotopic g.diagonal := by
+  rcases H with ⟨H⟩
+  refine ⟨H.compContinuousMap _, ?_⟩
+  rintro t _ (rfl | rfl)
+  · exact H.prop' _ _ Cube.zero_mem_boundary
+  · exact H.prop' _ _ Cube.one_mem_boundary
```

Comment:
This lemma doesn't seem to have existed in mathlib3 ... Why is it introduced without a porting note?